v4/opt/factory.go

```diff
@@ -245,6 +264,84 @@ func (f *Factory) appendColumnProjections(projections, group GroupID) GroupID {
 	return f.ConstructProjections(f.mem.storeList(items), f.mem.internPrivate(&outputCols))
 }
 
+// substitute recursively substitutes oldCol with newCol in the given filter
+// expression. For example, if oldCol=x and newCol=y, x < 5 will become y < 5.
+func (f *Factory) substitute(filter GroupID, oldCol, newCol ColumnIndex) GroupID {
```

Comment:
This looks correct to me. Maybe add a TODO to improve performance before we add into production code. In particular, it's important that if we don't find oldCol in a particular subtree, that we don't reconstruct the node. Ideally, we'd code generate some kind of "replace visitor" that would do the heavy lifting of rebuilding a tree bottom-up without creating unnecessary extra nodes.